docs/equality.md

```diff
@@ -0,0 +1,348 @@
+# Compiling Equality and Comparison
+
+This spec covers how equality is compiled and executed by the F# compiler and library, based mainly on the types involved in the equality operation after all inlining, type specialization and other optimizations have been applied.
+
+### What do we mean by an equality operation?
+
+This spec is about the semantics and performance of the following coding constructs
+
+* `a = b`
+* `a <> b`
+
+It is also about the semantics and performance of uses of the following `FSharp.Core` constructs.
+
+which, after inlining, generate code that contains an equality check at the specific `EQTYPE`
+* `HashIdentity.Structural<'T>`
+* `{Array,Seq,List}.contains` 
+* `{Array,Seq,List}.countBy` 
+* `{Array,Seq,List}.groupBy`
+* `{Array,Seq,List}.distinct` 
+* `{Array,Seq,List}.distinctBy` 
+* `{Array,Seq,List}.except` 
+
```

Comment:
Well, none of the operations above are supported in maps and sets.

Reply:
Which operations? :-O
Comparison isn't considered as a "sibling of equality" in this context, being potentially affected by the same phenomena w.r.t. to boxing, inlining and type specialization?

Reply:
I am not sure I follow. Operations in question are `countBy`, `groupBy`, `distinct`, `distinctBy`, `except` - unavailable in maps and sets.

What exactly do you want to be noted here?

Reply:
No - my comment had nothing to do with specific questions, it is rather this section of the doc (and the doc alltogether) not mentioning neither Set/Map nor comparison at all.

If it is fully out of scope - why
If not - why

(because the similar applies if I read things correctly - callchain ends up with GenericComparisonWithComparerIntrinsic doing boxing on both arguments)

Reply:
Well, comparison is out of scope because the doc is about equality. Comparison is indeed similar in many (although not all) ways but the doc is already super loaded cognitively and we don't want to add even more information here, at least for now.

More importantly, this all is subject to change in the upcoming weeks, so don't think of this as of a final version. Don and I are going through the related code, PRs, suggestions and so on and will be probably changing, adding and restructuring information here. For now this is mostly meant to be a point of reference in PRs.